Alex and Happy Part 2/article.md

```diff
@@ -0,0 +1,1372 @@
+
+A **non-terminal** is a description of how the grammar can group symbols together. It's an arbitrary identifier that's replaced by other terminals and non-terminals (terminals and non-terminals together are called **symbols**) to produce derivations. A sequence that doesn't match any non-terminal is considered a syntax error.
+In the example below, `myNonTerminal` is non-terminal.
+They will represent the internal nodes of the abstract syntax tree.
+
+A **production rule** is a rule describing how to "produce" a string that is syntactically valid according to the grammar. It joins terminals together to form a sentence.
+This is how it looks in Happy:
+
+```happy
+myNonTerminal
+  : bodyA1 bodyA2 ... { actionA }
+  | bodyB1 bodyB2 ... { actionB }
+  | ...
+```
+
+n.b.: The ellipses here are not correct Happy grammar. We use them to represent that more symbols may be present.
+
+* `body@#` where `@` is some letter and `#` some number means that the parser must parse each body in succession.
+* The pipe (`|`) represents an alternation (an _or_). The parser may parse any of the given bodies, accepting the one that matches. A body of a production may be empty, meaning that `myNonTerminal : { action }` is also accepted.
+* `action@` represents what the parser should do once it successfully parses that rule, such as how to interpret what it has parsed or how to build a data structure with it. Like in Alex, it may be any Haskell expression.
+
+In Happy, production rules may also have an optional type signature.
+For instance, the declaration of `myNonTerminal` could also have been written as such:
+
+```happy
+myNonTerminal :: { MyType }
+  : bodyA1 bodyA2 ... { actionA }
+  | bodyB1 bodyB2 ... { actionB }
+  | ...
+```
+
+In the example above, any Haskell type may replace `MyType`.
+
+As a concrete example, let's describe a non-terminal for arithmetic operations, the goal of which is to evaluate the operations.
+Here, the entire snippet contains a production rule forming the `arithmetic` non-terminal.
+Assume that `number`, `'('`, `')'`, `'+'`, `'-'`, `'*'`, and `'/'` are terminals.
+
+```happy
+arithmetic :: { Double }
+  : number                    { $1 }
+  | arithmetic '+' arithmetic { $1 + $3 }
+  | arithmetic '-' arithmetic { $1 - $3 }
+  | arithmetic '*' arithmetic { $1 * $3 }
+  | arithmetic '/' arithmetic { $1 / $3 }
+  | '(' arithmetic ')'        { $2 }
+  | '-' arithmetic            { negate $2 }
+```
+
+Each `$n` refers to a symbol in the corresponding production body, respectively.
+
+There are a few more steps to get this example running, such as dealing with operator precedence and associativity.
+Our goal is to provide arithmetic operations for MiniML as well, so we'll describe how to do it later in the tutorial. :)
+
+We'll also use the following notation to represent productions.
+This is not valid Happy code, but it's what it uses for debugging information.
+This notation is also closer to what's used in formal language theory.
```

Comment:
Small nitpick, I would drop this below the code sample since it is more of an aside.